src/schema.jl

```diff
@@ -237,6 +239,150 @@ function apply_schema(t::ConstantTerm, schema::Schema, Mod::Type)
     InterceptTerm{t.n==1}()
 end
 
+# general idea is once we hit a FunctionTerm, we need to continue to
+# apply_schema because there might be some child that is un-protected.  So we
+# enter the Protected context and recursively apply_schema, and when we
+# encounter `unprotect` we restore the old context and continue to recursively
+# apply_schema
+"""
+    struct Protected{Ctx}
+
+Represent a context in which the normal special syntax and
+[`apply_schema`](@ref) transformations should not apply.  This is automatically
+applied to the arguments of a [`FunctionTerm`](@ref), meaning that by default
+calls to `+`, `&`, or `~` inside a [`FunctionTerm`](@ref) will be interpreted as
+calls to the normal Julia functions, rather than term union, interaction, or
+formula separation.
+
+The only special behavior with [`apply_schema`](@ref) inside a `Protected`
+context is when a call to [`unprotect`](@ref) is encountered.  At that point,
+everything below the call to `unprotect` is treated as special formula syntax.
+
+A `Protected` context is created inside a [`FunctionTerm`](@ref) automatically,
+but can be manually created with a call to [`protect`](@ref).
+```
+"""
+struct Protected{Ctx} end
+Base.broadcastable(x::Protected) = Ref(x)
+"""
+    protect(term::T)
+
+Create a [`Protected`](@ref) context for interpreting `term` (and descendents) during
+`apply_schema`.
+
+Outside a [`@formula`](@ref), acts as a constructor for the singleton `Protected{T}`.
+
+# Example
+
+```jldoctest; setup = :(using Random; Random.seed!(1))
+julia> d = (y=rand(4), a=[1:4;], b=rand(4))
+
+julia> f = @formula(y ~ 1 + protect(a+b));
+
+julia> modelmatrix(f.rhs, d)
+4×2 Array{Float64,2}:
+ 1.0  1.48861
+ 1.0  2.21097
+ 1.0  3.95192
+ 1.0  4.9999
+
+julia> d.a .+ d.b
+4-element Array{Float64,1}:
+ 1.4886128300795012
+ 2.210968202158536
+ 3.951916339835734
+ 4.999904658898614
+```
+"""
+protect(ctx) = Protected{ctx}()
+# construct singletons to avoid method ambiguities using Type{<:Protected}
```

Comment:
I don't understand to what code this comment applies.

Reply:
the bit above.  `protect(ctx)` _could_ return `Protected{ctx}` but that leads to ambiguities using `Type{<:Protected}` for dispatch instaed of `Protected` for some reason I've now forgotten

Reply:
OK. Then maybe this would be more explicit?
```suggestion
# return instances rather than types to avoid method ambiguities using Type{<:Protected}
```